qcodes/dataset/data_set_cache.py

```diff
@@ -0,0 +1,102 @@
+from typing import TYPE_CHECKING, Dict, Optional
+
+import numpy as np
+
+from qcodes.dataset.sqlite.queries import (
+    get_interdeps_from_result_table_name, completed,
+    get_parameter_data_for_one_paramtree)
+
+if TYPE_CHECKING:
+    import pandas as pd
+    from .data_set import DataSet, ParameterData
+
+
+class DataSetCache:
+    """
+    The DataSetCache contains a in memory representation of the
+    data in this dataset as well a a method to progressively read data
+    from the db as it is written. The cache is available in the same formats
+    as :py:class:`.DataSet.get_parameter_data` and :py:class:`.DataSet.get_data_as_pandas_dataframe`
+
+    """
+
+    def __init__(self, dataset: 'DataSet'):
+        self._dataset = dataset
+        self._data: ParameterData = {}
+        #: number of rows read per parameter tree (by the name of the dependent parameter)
+        self._read_status: Dict[str, int] = {}
+        self._loaded_from_completed_ds = False
+
+    def load_data_from_db(self) -> None:
+        """
+        Loads data from the dataset into the cache.
+        If new data has been added to the dataset since the last time 
+        this method was called, calling this method again would load 
+        that new portion of the data and append to the already loaded data.
+        If the dataset is marked completed and data has already been loaded
+        no load will be performed.
+        """
+        if self._loaded_from_completed_ds:
+            return
```

Comment:
wow, codecov tells that we don't have a test for this case :)

Reply:
Yes, rewrote the tests to ensure that we also trigger this